src/txn/txn.c

```diff
@@ -630,39 +630,50 @@ __txn_commit_timestamp_validate(WT_SESSION_IMPL *session)
 		WT_RET_MSG(session, EINVAL, "no commit_timestamp required and "
 		    "timestamp set on this transaction");
 
-	if (WT_VERBOSE_ISSET(session, WT_VERB_TIMESTAMP)) {
-		/*
-		 * Error on any valid update structures for the same key that
-		 * are at a later timestamp.
-		 */
-		for (i = 0, op = txn->mod; i < txn->mod_count; i++, op++) {
-			if (op->type != WT_TXN_OP_BASIC_TS)
-				continue;
+	/*
+	 * If we're not doing any key consistency checking, we're done.
+	 */
+	if (!F_ISSET(txn, WT_TXN_TS_COMMIT_KEYS))
+		return (0);
+
+	/*
+	 * Error on any valid update structures for the same key that
+	 * are at a later timestamp or use timestamps inconsistently.
+	 */
+	for (i = 0, op = txn->mod; i < txn->mod_count; i++, op++)
+		if (op->type == WT_TXN_OP_BASIC_TS ||
+		    op->type == WT_TXN_OP_BASIC) {
 			/*
 			 * Skip over any aborted update structures.
 			 */
 			upd = op->u.upd->next;
 			while (upd != NULL && upd->txnid == WT_TXN_ABORTED)
 				upd = upd->next;
+
 			/*
 			 * Check the timestamp on this update with the
 			 * first valid update in the chain. They're in
 			 * most recent order.
 			 */
-			if (upd != NULL &&
-			    __wt_timestamp_cmp(&op->u.upd->timestamp,
-			    &upd->timestamp) < 0) {
-				WT_RET(__wt_timestamp_to_hex_string(session,
-				    timestamp_buf[0], &op->u.upd->timestamp));
-				WT_RET(__wt_timestamp_to_hex_string(session,
-				    timestamp_buf[1], &upd->timestamp));
-				__wt_verbose(session, WT_VERB_TIMESTAMP,
-				    "Timestamp %s on new update is older than "
-				    "timestamp %s on existing update.",
-				    timestamp_buf[0], timestamp_buf[1]);
-			}
+			if (upd == NULL)
+				continue;
+			/*
+			 * Check for consistent per-key timestamp usage.
+			 * If timestamps are or are not used originally then
+			 * they should be used the same way always. Check
+			 * timestamps are used in order.
+			 */
+			op_used_ts =
+			    __wt_timestamp_iszero(&op->u.upd->timestamp);
+			upd_used_ts = __wt_timestamp_iszero(&upd->timestamp);
```

Comment:
I thought maybe there was a case that went from not using timestamps to using timestamps (once)? At the end of initial sync in MongoDB.

Reply:
I believe there is a table that goes from not using timestamps to using them for all keys. @dgottlieb is that correct? This config is primarily for the mdb catalog where some keys always have timestamps and some keys never have timestamps so the broader per-table settings are not able to be used.

Reply:
I fear I've written more than what you were both asking about. If either of you have further clarifying questions, I'll try to give a more focused response. :)

> I thought maybe there was a case that went from not using timestamps to using timestamps (once)? At the end of initial sync in MongoDB.

That is generally correct. The more nuanced story is that initial sync has a few phases. The first cloning phase I believe will never timestamp any writes. Then there is an oplog application phase which will timestamp writes. After that is an optional "document fetching" phase which will not timestamp writes. For now, my goal for verification is to not check anything during the entirety of initial sync. We don't support reading from a node at a time when it was initial syncing so it's less important. If there was interest, I believe there is a path to appropriate alter the table settings between each phase.

What I want to do when MongoDB comes out of initial sync, is have storage set each table with its appropriate verification setting. I suspect iterating through each table may be a penalty we don't want to pay in general (think customers with thousands of tables), but I think it's important to do this at least in our evergreen debug build tests.

> I believe there is a table that goes from not using timestamps to using them for all keys.

There are two cases I know of that mostly* match this description. The first being the initial sync process @agorrod mentioned. The second is for index builds. Index builds on an existing collection can be updated in two ways. One is the collection scan to initialize the index. The other is when a user performs a concurrent insert/update/delete to a collection while the index is building. The first is not timestamped, the second is. Foreground index builds don't allow concurrent updates during the build process. Those tables should experience a sudden change in timestamps as @sueloverso mentioned. Background index builds do allow concurrent updates and will have a mix of both varieties. Notably this can include a timestamped write to an index entry followed by a non-timestamped write.

Similarly to initial sync, I don't plan on enforcing any timestamp behavior on indexes until they are finished building.

The purpose of this ticket, as @sueloverso described, is for the `_mdb_catalog`. Once again, I won't enforce any timestamping until initial sync completes. After the node is in steady-state replication, this table has the unique property of containing data for replicated and non-replicated collections. To keep parity with existing rollback behavior, we don't want to rollback updates to non-replicated collections. 

Accidentally timestamping some writes to non-replicated metadata documents can cause a `rollback_to_stable` or, more simply, a read transaction at a timestamp to return an incorrect result (that is, if MongoDB's high-level catalog allowed for versioned reads). Likewise, accidentally forgetting to timestamp a write to a replicated collection's metadata document can result in similar corruption.

The work to squash out any remaining (known) untimestamped writes is nearing completion (tickets linked from https://jira.mongodb.org/browse/SERVER-30809 ). All of those tickets have to do with timestamping writes to the `_mdb_catalog`, which is why I'm particularly interested in having independent validation that we got it right.

Reply:
Thanks for the explanation @dgottlieb that all makes sense to me - I'll go ahead and merge.